libsrc/apple2/get_tv.s

```diff
@@ -0,0 +1,151 @@
+.endproc
+
+; Carry set at enter: wait for VBL +
+; Carry clear at enter: wait for VBL -
+; Increments X every 85 cycles.
+.proc count_until_vbl_bit
+        lda     #$10          ; BPL
+        bcc     :+
+        lda     #$30          ; BMI
+:       sta     sign
+
+        bit     PTRIG         ; Slow accelerators hopefully
+
+        ; Wait for VBLsign change with 85 cycles loops
+:       nop                   ; 2
+        inx                   ; 4
+        jsr     waste_72      ; 78
+        bit     RDVBLBAR      ; 82 - Wait for VBL change
+sign:
+        bpl     :-            ; 85 - patched with bpl/bmi
+        rts
+.endproc
+
+.proc calibrate_iic
+        php
+        sei
+
+        sta     IOUDISOFF
+        lda     RDVBLMSK
+        pha                   ; Back up for cleanup
+
+        bit     ENVBL
+        bit     PTRIG         ; Reset VBL interrupt flag
+:       bit     RDVBLBAR      ; Wait for one VBL
+        bpl     :-
+
+        ldx     #$00
+        clc
+        jsr     count_until_vbl_bit
+
+        pla                   ; Cleanup
+        asl
+        bcs     :+            ; VBL interrupts were already enabled
+        bit     DISVBL
+:       sta     IOUDISON      ; IIc Tech Ref Man: The firmware normally leaves IOUDIS on.
+
+        plp
+        jmp     vbl_calibrate_done
+.endproc
+
+.proc calibrate_iie
+:       bit     RDVBLBAR      ; Wait for bit 7 to be off (VBL start)
+        bmi     :-
+:       bit     RDVBLBAR      ; Wait for bit 7 to be on (VBL end)
+        bpl     :-
+
+        ; Wait and count during a full cycle
+        ldx     #0
+        sec
+        jsr     count_until_vbl_bit
+        clc
+        jsr     count_until_vbl_bit
+
+        jmp     vbl_calibrate_done
+.endproc
+
+.proc calibrate_iigs
+        ; Backup speed and slow down
+        jsr     _get_iigs_speed
+        pha
+        lda     #SPEED_SLOW
+        jsr     _set_iigs_speed
+
+        ; The same as IIe, but reverted, because fuck you that's why
+:       bit     RDVBLBAR      ; Wait for bit 7 to be on (VBL start)
+        bpl     :-
+:       bit     RDVBLBAR      ; Wait for bit 7 to be off (VBL end)
+        bmi     :-
+
+        ; Wait and count during a full cycle
+        ldx     #0
+        clc
+        jsr     count_until_vbl_bit
+        sec
+        jsr     count_until_vbl_bit
+
+        jsr     vbl_calibrate_done
+
+        ; Restore user speed
+        pla
+        jmp     _set_iigs_speed
+.endproc
+
+.proc vbl_calibrate_done
+        lda     #TV_NTSC      ; Consider we're at 60Hz
+        sta     _get_tv+1
+        ; we'll now have X ~= 205 at 60Hz, ~= 240 at 50Hz:
+        ; 205*85 = 17340, 240*85 = 20400.
+        ; Consider any X >= 220 means 50Hz
+        cpx     #220
+        bcc     :+
+        lda     #TV_PAL       ; we're at 50Hz
+        sta     _get_tv+1
+:       rts
+.endproc
+
+; The only thing remaining from that code after init
+.segment "CODE"
+
+.proc _get_tv
+        lda     #TV_OTHER
```

Comment:
Hm, not sure if self-modifying code is a bit over-engineered here. I.e. https://github.com/cc65/cc65/blob/master/libsrc/apple2/get_ostype.s doesn't use it which causes two more bytes. If you feel that it is desirable I ask you to modify get_ostype (in another PR) as well for consistency reasons.

Reply:
I don't think that's important, it's better if it's clean. Having _get_tv or _get_ostype in DATA feels weird :) I've pushed a change to mimic ostype (with tv in INIT).